### ISSUE TITLE
Optional API for challenge-based R1CS proofs

### ISSUE THREAD

**Issue by oleganza** [dalek-cryptography/bulletproofs#242]
(This is a summary of an issue raised by @hdevalence in an offline discussion.)

The challenge-based constraints in R1CS API (#222) are pretty novel and not have a formal security proof yet. It makes sense to make the use of those _optional_ with respect to the rest of the R1CS API.

### Proposal

First, it should be possible for the verifier to ensure that randomization API is not available throughout the hierarchy of gadgets unless it was explicitly turned on. This makes challenge-based circuits an "opt-in" feature.

The proofs for non-randomized constraints should be compatible with the proofs for randomizable constraints: that is, they could be verified by a less conservative verifier, as well as a conservative one. For this we'd simply need to pad the transcript with dummy zero commitments `(A_I2, A_O2, S2)`. This padding should obviously not affect the correctness of the original R1CS protocol.

### Questions

1. Do we have different proof types for with randomization and without randomization; or do we use the uniform type with a flag distinguishing the two?
2. Do we try to save 96 bytes of all-zero second-phase commitments `(A_I2, A_O2, S2)` for the proofs of non-randomized CS, or just have a uniform proof format?
3. Do we turn on/off the randomization API dynamically or statically? E.g. is a subtrait of a `RandomizableConstraintSystem: ConstraintSystem`? Gadgets that do not use randomization could use a more conservative supertrait, but gadgets that need it would use a subtrait.

WDYT? @hdevalence @cathieyun

**Comment by oleganza** [comment]
The minimal support for compact proofs would be:

1. One type `R1CSProof` as we have right now.
2. When serializing, check if the 3 second-phase commitments are `.is_identity()`, then avoid encoding them altogether.
3. When deserializing, check if the byteslice is odd - in this case replace missing 3 points with 3 identity points.
4. When making a proof, make sure these commitments are left as identity points (with zero blinding factors) in case there are no second-phase variable allocations.

**Comment by hdevalence** [comment]
> The prover/verifier knows whether they are going to use randomization (that's baked into the circuit they use), but they know it _late_: when the gadgets actually trigger randomization. @hdevalence's suggestion is to look at the size of the proof bytestring and if it's even, it's a 2-phase proof, if it's odd - it's a 1-phase proof (16+2k elements VS 13+2k elements). This way we can avoid having a special flag.

Just a little bit of clarification, the _late_ here isn't about the proving / verification logic, but about the fact that verification doesn't work on raw bytes, but on an `R1CSProof` struct.

Since the state transition is roughly 
```text
|init CS| -> add constraints -> |finalized CS| -> proving -> |proof|
|init CS| -> add constraints -> |finalized CS| -> verification -> |is ok?|
```
the prover and verifier both know whether randomization is in use, but the problem is that the `R1CSProof` struct is parsed from bytes independently of the constraint system. But the serialized proof is either `fixed number + even number` of 32-byte elements or `fixed number + even number + 3` 32-byte elements, so the parity of `len/32` can distinguish the two types.

When proving or verifying, the constraint system should add a 2-phase domain separator when randomized commitments are used, to ensure that the two cases can't be mixed.

**Comment by oleganza** [comment]
@hdevalence what's the best place to add a domain separator? Seems like it should be at the beginning of the 2nd phase: if the list of callbacks is empty, we commit "1-phase" prefix (and the proof is shorter) and if it's non-empty, we commit "2-phase" prefix (and the proof is longer).

**Comment by hdevalence** [comment]
Yeah, that seems right to me, adding a `b"dom-sep", b"r1cs-1phase"` or `b"dom-sep", b"r1cs-2phase"` to the transcript at that point seems like it should be the right place, so that all the 2-phase challenges are bound to the dom-sep

**Comment by oleganza** [comment]
Interestingly, the size of the proof is not fully aligned with whether randomized constraints were used: e.g. it's possible to squeeze challenges and add some linear constraints with them, but not allocate any more variables. Luckily, we can treat these things independently:

1. No second phase allocations <=> shorter proof
2. No second phase callbacks <=> no extra domain separator

**Comment by hdevalence** [comment]
Adding more linear constraints doesn't change the size of the R1CS part of the proof, only the lengths of the inner product (and hence its size), right? But the inner-product proof always has an even number of points, so I don't think I really understand if/how this changes the even/odd criteria.

**Comment by oleganza** [comment]
>Adding more linear constraints doesn't change the size of the R1CS part of the proof, only the lengths of the inner product (and hence its size), right?

Not even IPP — more linear constraints without extra allocations do not change the size of the proof at all. The even/odd criteria is not changed.

What I mean to say: you can have these combinations of things:

1. No multipliers and no linear constraints in the second phase (odd proof, domain separator is "phase1").
2. Some multipliers and some linear constraints in the second phase (even proof, domain separator is "phase2").  
3. No multipliers in the second phase, but some linear constraints added (odd proof, domain separator is "phase2").
4. Unrealistic case: have some multipliers in 2nd phase, but none of linear constraints — means you are not really using challenges (even proof, domain separator is "phase1").

**Comment by oleganza** [comment]
Closed by #264
